### PR TITLE
git-hack and git-ship only fetch and push if remote repo exists

### LINKED ISSUE #370: git hack fails without remote
When I create a local repo, and do not have a shared repo, git hack fails. The branch has not been created, and I am still on the master branch.

I'd either expect it to work locally, or tell me that git town requires a remote repo. 

```
reto ~/AndroidStudioProjects/Quakey>master>$ git hack phidget
[master] git fetch --prune
fatal: No remote repository specified.  Please, specify either a URL or a
remote name from which new revisions should be fetched.

[master] git stash -u
Saved working directory and index state WIP on master: 8565176 Added place name feedback, removed unnecessary options
HEAD is now at 8565176 Added place name feedback, removed unnecessary options

[master] git push -u origin master
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

To abort, run "git hack --abort".
To continue after you have resolved the conflicts, run "git hack --continue".

reto ~/AndroidStudioProjects/Quakey>master>$ 
```

### PR DESCRIPTION
**Pull request by ricmatsui**
If the repo has a remote url, then do as before, otherwise skip any remote operations. Closes #370 

This definitely came up when I wanted to just try it out at first.

### CODE REVIEW

File: features/git-hack/on_main_branch/without_remote_origin/with_open_changes.feature [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  Scenario: result
+    Then it runs the Git commands
+      | BRANCH      | COMMAND                          |
+      | main        | git stash -u                     |
+      | main        | git checkout -b new_feature main |
+      | new_feature | git stash pop                    |
+    And I end up on the "new_feature" branch
+    And I still have an uncommitted file with name: "uncommitted" and content: "stuff"
+    And the branch "new_feature" has not been pushed to the repository
```
Comment (kevgo) [inline features/git-hack/on_main_branch/without_remote_origin/with_open_changes.feature]: I would say that given that this feature spec explicitly declares that there is no repository, and the Git commands show no push, this particular step could be omitted. Otherwise it is confusing ("Which repository is the branch not pushed to? I thought there is no repository?").

File: features/git-hack/on_main_branch/without_remote_origin/without_open_changes.feature [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    And I am on the "main" branch
+    When I run `git hack new_feature`
+
+
+  Scenario: result
+    Then it runs the Git commands
+      | BRANCH | COMMAND                          |
+      | main   | git checkout -b new_feature main |
+    And I end up on the "new_feature" branch
+    And the branch "new_feature" has not been pushed to the repository
```
Comment (kevgo) [inline features/git-hack/on_main_branch/without_remote_origin/without_open_changes.feature]: ditto, I propose removing this step

File: features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  Scenario: result
+    Then it runs the Git commands
+      | BRANCH  | COMMAND                            |
+      | feature | git checkout main                  |
+      | main    | git checkout feature               |
+      | feature | git merge --no-edit main           |
+      | feature | git checkout main                  |
+      | main    | git merge --squash feature         |
+      | main    | git commit -m "feature done"       |
```
Comment (kevgo) [inline features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature]: Please fix the indentation of this table by running `rake format`.

File: src/git-ship [hunk 4]
```diff
@@ -27,6 +27,8 @@ function preconditions {
 
 
 function steps {
+  local has_remote="$(has_remote_url)"
```
Comment (kevgo) [inline src/git-ship]: Cool idea to cache this here!

**Comment by kevgo** [comment]
This is an awesome change, @ricmatsui, thanks so much for helping! This is very clean and well done, nice work!

If you run the specs as `rake`, it also runs the linters (we lint both the Ruby and the Shell code using [Rubocop](https://github.com/bbatsov/rubocop) and [ShellCheck](http://www.shellcheck.net)), and then it runs the specs in parallel, i.e. faster. If you run `rake lint`, it runs just the linters. Please make sure they also pass. 

Apart from the few cosmetic changes mentioned above, this looks good to go from my end. Feature specs are green. What do you think, @charlierudolph and @allewun?

File: src/helpers/git_helpers/remote_helpers.sh [hunk 5]
```diff
@@ -18,3 +18,12 @@ function remote_repository_name {
   local domain=$(remote_domain)
   remote_url | sed -E "s#.*$domain[/:](.+)#\1#" | sed "s/\.git$//"
 }
+
+# Returns true if a remote repository exists
+function has_remote_url {
+  if [ ! "$(remote_url)" ]; then
```
Comment (charlierudolph) [inline src/helpers/git_helpers/remote_helpers.sh]: I'd prefer the following (more consistent with the rest of the checks throughout the codebase)
```bash
if [ -z "$(remote_url)"]; then
```
`-z` checks if a string is non-empty

File: src/helpers/git_helpers/remote_helpers.sh [hunk 6]
```diff
@@ -18,3 +18,12 @@ function remote_repository_name {
   local domain=$(remote_domain)
   remote_url | sed -E "s#.*$domain[/:](.+)#\1#" | sed "s/\.git$//"
 }
+
```
Comment (charlierudolph) [inline src/helpers/git_helpers/remote_helpers.sh]: Please add one more empty line (the codebase is following the pattern of two empty lines between functions)

File: src/helpers/git_helpers/remote_helpers.sh [hunk 7]
```diff
@@ -18,3 +18,12 @@ function remote_repository_name {
...
+# Returns true if a remote repository exists
```
Comment (charlierudolph) [inline src/helpers/git_helpers/remote_helpers.sh]:
```
# Returns true if the repository has a remote configured
```

File: src/git-ship [hunk 8]
```diff
@@ -34,9 +36,17 @@ function steps {
   fi
 
   echo "checkout_main_branch"
-  echo "fetch"
+
```
Comment (charlierudolph) [inline src/git-ship]: Please remove the added newline. These two steps are closely related so I'd prefer keeping this right next to each other. Same for the other newlines added before the other `if [ "$has_remote" = true ]; then` statements

File: src/git-hack [hunk 9]
```diff
@@ -26,7 +30,11 @@ function steps {
 
   echo "checkout_main_branch"
   echo "rebase_tracking_branch"
-  echo "push"
+
```
Comment (charlierudolph) [inline src/git-hack]: Please remove this empty line. This step is closely related to the other two

**Comment by charlierudolph** [comment]
Thanks @ricmatsui. @kevgo this was actually something I brought up early on #89. You had suggested doing exactly what @ricmatsui implemented, but we ended up pushing it off as unimportant. I believe we should add this support for the other commands as well in separate PRs.

**Comment by kevgo** [comment]
I wouldn't say unimportant. Given that we always use remotes here, we decided to wait until somebody actually needs this feature. Now that this has happened by several people, yes, let's make sure all commands work with local-only repos!  :)

File: src/git-hack (hunk 9 above)
Comment (allewun) [inline src/git-hack]: True but I think this is nicer to read since the if-block takes up 3 lines. Perhaps we can add an `echo_if_true($1=string, $2=conditional)` for concision.

Comment (charlierudolph) [inline src/git-hack]: :+1: I like it

Comment (kevgo) [inline src/git-hack]: Good idea. Let's do that in a separate pull request. This is good as is then.

File: src/git-ship (hunk 8 above)
Comment (charlierudolph) [inline src/git-ship]: Good as is. Will be dealt with separately

File: features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      | BRANCH  | LOCATION | FILE NAME    | FILE CONTENT    |
+      | feature | local    | feature_file | feature content |
+    And I am on the "feature" branch
+    When I run `git ship -m "feature done"`
+
+
+  Scenario: result
+    Then it runs the Git commands
+      | BRANCH  | COMMAND                            |
+      | feature | git checkout main                  |
```
Comment (allewun) [inline features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature]: Can you see if you can save `git-ship` from performing this extraneous checkout? Same optimization for `git-hack` can be made as well.

**Comment by allewun** [comment]
Cool stuff @ricmatsui! Great job keeping things clean and consistent with the codebase. Looks good to go from my end once the comments have been addressed.

I've created issue #404 for adding some style guidelines to our contribution doc, which should help future contributors like yourself.

File: features/git-hack/on_main_branch/without_remote_origin/with_open_changes.feature (hunk 1 above)
Comment (ricmatsui) [inline features/git-hack/on_main_branch/without_remote_origin/with_open_changes.feature]: Agreed, removed.

File: features/git-hack/on_main_branch/without_remote_origin/without_open_changes.feature (hunk 2 above)
Comment (ricmatsui) [inline features/git-hack/on_main_branch/without_remote_origin/without_open_changes.feature]: Agreed, removed.

File: features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature (hunk 3 above)
Comment (ricmatsui) [inline features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature]: Fixed

File: src/git-ship (hunk 4 above)
Comment (ricmatsui) [inline src/git-ship]: Thanks!

File: src/helpers/git_helpers/remote_helpers.sh (hunk 5 above)
Comment (ricmatsui) [inline src/helpers/git_helpers/remote_helpers.sh]: No problem.

File: src/helpers/git_helpers/remote_helpers.sh (hunk 6 above)
Comment (ricmatsui) [inline src/helpers/git_helpers/remote_helpers.sh]: Yeah, no problem.

File: src/helpers/git_helpers/remote_helpers.sh (hunk 7 above)
Comment (ricmatsui) [inline src/helpers/git_helpers/remote_helpers.sh]: Done

File: src/git-hack (hunk 9 above)
Comment (ricmatsui) [inline src/git-hack]: Sounds good.

File: features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature (hunk 10 above)
Comment (ricmatsui) [inline features/git-ship/current_branch/on_feature_branch/without_open_changes/without_remote_origin.feature]: Yes, you're right. I misunderstood how the new branches are made, so these extraneous checkouts are not made anymore.

**Comment by ricmatsui** [comment]
@kevgo Thanks! I fixed the linter issues, and I also had to touch the configuration and environment helpers.
@allewun Sounds good! Yes, I tried inferring as much as I could.

All the comments should be addressed now.

**Comment by kevgo** [comment]
Excellent work, @ricmatsui! If I don't hear any more comments from the other guys, I'll merge this PR tomorrow. To set this up, can you please rebase this change so that it is a single commit again?

**Comment by ricmatsui** [comment]
Awesome, thanks! I've rebased it now.

**Comment by kevgo** [comment]
Merged. Thanks @ricmatsui, this was an really well done pull request! How do you feel about applying the same improvement to `git kill` and `git sync`? I think they also make sense without a remote, and this should be a pretty straightforward change now.

**Comment by ricmatsui** [comment]
Yes, that's the plan. I'll probably start that today.
